### PR TITLE
build: clean up the VERSION file

### LINKED ISSUE #2844: Conflict between LLVM's version file and our VERSION file
On MacOS, which generally has case-insensitive file systems, the `VERSION` file that gets generated while building Charm++ conflicts with a `version` header file provided by LLVM, leading to compilation errors:

```
clang++ -fPIC -dynamic -fno-common -stdlib=libc++ -DCMK_GFORTRAN -mmacosx-version-min=10.7 -D_DARWIN_C_SOURCE -I../include -I. -c test.cpp -o test.o -std=c++11
In file included from test.cpp:3:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/map:479:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/__tree:14:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/iterator:421:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/__functional_base:14:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/type_traits:417:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/cstddef:37:
../include/version:1:1: error: expected unqualified-id
61001
^
```

See https://github.com/llvm-mirror/libcxx/blob/master/include/cstddef#L37

This is with llvm 10.0 (homebrew) on MacOS 10.15.4, both old and new build systems are affected.
llvm 9 is probably affected too.

Potential fixes:
- rename our `VERSION` file.
- make our `VERSION` file C-compliant (e.g. by adding `//` before the version number). This would avoid the compilation error, but LLVM's version file would still not be included (which is not a critical error I think).
- remove `VERSION` completely.

### PR DESCRIPTION
**Pull request by matthiasdiener**
create two separate files, `charm-version.h` and `charm-version-git.h` in
`include/`, with the release/API version and git version respectively.

The files look like this:

```
$ cat ./netlrts-darwin-x86_64/include/charm-version.h
#define CHARM_VERSION 61002
#define CHARM_VERSION_MAJOR 6
#define CHARM_VERSION_MINOR 10
#define CHARM_VERSION_PATCH 2
$ cat ./netlrts-darwin-x86_64/include/charm-version-git.h
#define CHARM_VERSION_GIT "v6.11.0-devel-430-g50bd748ad*"
```

Fixes #2844.

For the old build system, `commitid.sh` creates the `charm-version-git.h` file containing the git version of Charm++ (this version was previously contained in `tmp/VERSION`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1095,10 +1095,11 @@ add_custom_target(create_symlinks ALL
 #   add_dependencies(create_symlinks createlink)
 # endif()
 
-# Create the .vdir/.gdir/VERSION files
+# Create the .vdir/.gdir/charm-version.h/charm-api-version.h files
 file(WRITE ${CMAKE_BINARY_DIR}/include/.gdir   "${CMK_GDIR}\n")
 file(WRITE ${CMAKE_BINARY_DIR}/include/.vdir   "${CMK_VDIR}\n")
-file(WRITE ${CMAKE_BINARY_DIR}/include/VERSION "${CHARM_VERSION}\n")
+file(WRITE ${CMAKE_BINARY_DIR}/include/charm-version.h "#define CHARM_VERSION "${CHARM_VERSION}"\n")
+file(WRITE ${CMAKE_BINARY_DIR}/include/charm-api-version.h "#define CHARM_API_VERSION ${PROJECT_VERSION}\n")
```
Comment (nilsleiffischer) [inline CMakeLists.txt:1102]: I suggest to `#define CHARM_VERSION_{MAJOR,MINOR,PATCH}` as well

**Comment by matthiasdiener** [comment]
This is somewhat risky and the changes are not very well tested, but it fixes an important issue users have been seeing. @rbuch: Could you please check if the changes related to the tarball packaging work as expected?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1095,10 +1095,11 @@ add_custom_target(create_symlinks ALL
 #   add_dependencies(create_symlinks createlink)
 # endif()
 
-# Create the .vdir/.gdir/VERSION files
+# Create the .vdir/.gdir/charm-version.h/charm-api-version.h files
 file(WRITE ${CMAKE_BINARY_DIR}/include/.gdir   "${CMK_GDIR}\n")
 file(WRITE ${CMAKE_BINARY_DIR}/include/.vdir   "${CMK_VDIR}\n")
-file(WRITE ${CMAKE_BINARY_DIR}/include/VERSION "${CHARM_VERSION}\n")
+file(WRITE ${CMAKE_BINARY_DIR}/include/charm-version.h "#define CHARM_VERSION "${CHARM_VERSION}"\n")
```
Comment (nilsleiffischer) [inline CMakeLists.txt:1101]: The `"` need to be escaped, and the line below also needs the quotation marks

Comment (matthiasdiener) [reply CMakeLists.txt]: Fixed, thanks!

File: src/scripts/charmc [hunk 3]
```diff
@@ -389,7 +389,7 @@ filterLegacyLB()
 printVersion()
 {
 	findCharmBin $0
-	version=`cat $CHARMBIN/../include/VERSION`
+	version=`cat $CHARMBIN/../include/charm-version.h`
```
Comment (rbuch) [inline src/scripts/charmc:392]: This prints out the entire define line.

File: package-tarball.sh [hunk 4]
```diff
@@ -65,7 +65,7 @@ git add -u
 object_id=$(git write-tree)
 
 # Construct the target file/folder name
-version="charm-$(cat src/scripts/VERSION)"
+version="charm-$(cat src/scripts/charm-version.h)"
```
Comment (rbuch) [inline package-tarball.sh:68]: This prints out the entire define line.

**Review by rbuch — COMMENTED** [review]
Using `$(cat charm-version.h | grep CHARM_VERSION | awk '{print $3}')` in `package-tarball.sh` doesn't strip the version of the quotation marks, is there some other command that should be used for that? I'm confused why its use in `src/scripts/Makefile` doesn't result in extra quotation marks.

File: src/scripts/charmc (hunk 3 above)
Comment (matthiasdiener) [reply src/scripts/charmc]: Fixed, thanks. With the most recent commit in this PR, this prints the API version, not the "git" version (as previously). f you want, we can change that.

File: package-tarball.sh (hunk 4 above)
Comment (matthiasdiener) [reply package-tarball.sh]: fixed, thanks.

**Comment by matthiasdiener** [comment]
> Using `$(cat charm-version.h | grep CHARM_VERSION | awk '{print $3}')` in `package-tarball.sh` doesn't strip the version of the quotation marks, is there some other command that should be used for that? I'm confused why its use in `src/scripts/Makefile` doesn't result in extra quotation marks.

Good catch. Both `package-tarball.sh` and `src/scripts/Makefile` were incorrect and should be fixed now.

**Comment by evan-charmworks** [comment]
Should this PR add mention of `VERSION`'s removal to `CHANGES`? It might be good to offer documentation on the forward transition path for any Charm users who parse it.

**Comment by matthiasdiener** [comment]
> Should this PR add mention of `VERSION`'s removal to `CHANGES`? It might be good to offer documentation on the forward transition path for any Charm users who parse it.

We could add this and also add documentation to the manual regarding `CHARM_API_VERSION` etc.
Another idea is to `#include` both files into `charm-api.h` or so.

File: src/scripts/charmc (hunk 3 above)
Comment (rbuch) [reply src/scripts/charmc]: That sounds fine to me, so long as the preamble when running a program still prints out the "git" version.

**Comment by matthiasdiener** [comment]
Turns out that CHARM_VERSION is already used in conv-autoconfig.h (value is the same as CHARM_API_VERSION).
So I'll rename CHARM_API_VERSION_MAJOR etc. to CHARM_VERSION_MAJOR, and CHARM_VERSION to CHARM_VERSION_GIT.

**Comment by rbuch** [comment]
One thing to note is some applications (at least Enzo-E) read the root VERSION file as part of their build process. I don't think that should stop us from making this change, just know that it will be a breaking one.

File: src/scripts/configure.ac [hunk 5]
```diff
@@ -42,15 +42,26 @@ CHARMINC="."
...
+echo "#define CHARM_VERSION_PATCH $CHARM_VERSION_PATCH" >> ../include/charm-version.h
 
 charmout="charmconfig.out"
 /bin/rm -rf $charmout
 MAKEFILE_EXT=Make.extlib
 
 Echo() {
-	echo $* 
+	echo $*
 	echo $* >> $charmout
```
Comment (matthiasdiener) [inline src/scripts/configure.ac:65]:
```suggestion
	echo $* 
	echo $* >> $charmout
```

File: src/ck-core/charm-api.h [hunk 6]
```diff
@@ -39,10 +42,10 @@ libraries written in Charm for other languages.
 #endif
 
 
-/** 
+/**
```
Comment (matthiasdiener) [inline src/ck-core/charm-api.h:45]:
```suggestion
/** 
```

File: src/ck-core/charm-api.h [hunk 7]
```diff
@@ -39,10 +42,10 @@ libraries written in Charm for other languages.
...
  * Define a new Fortran-callable API routine, returning void,
- * that does nothing but map its arguments onto some 
- * (presumably analogous) C routine. 
+ * that does nothing but map its arguments onto some
+ * (presumably analogous) C routine.
```
Comment (matthiasdiener) [inline src/ck-core/charm-api.h:48]:
```suggestion
 * that does nothing but map its arguments onto some 
 * (presumably analogous) C routine. 
```

File: src/ck-core/charm-api.h [hunk 8]
```diff
@@ -56,7 +59,7 @@ FTN_NAME(CAPITALNAME,lowername) routine_args { \
 	Cname c_args;\
 }
 
-/** 
+/**
```
Comment (matthiasdiener) [inline src/ck-core/charm-api.h:62]:
```suggestion
/** 
```

**Comment by matthiasdiener** [comment]
Ready for review.

**Comment by evan-charmworks** [comment]
Does this work correctly from a packaged tarball too? I get this from CMake with current master from a tarball:

```
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
==============================
-- Configuring done
Charm++ 6.10.2 configuration: 
  OS version:      Ubuntu 18.04.5 LTS (Linux 5.4.0-1031-azure, x86_64)
  Charm++ version:  (dev)
```

I don't know if passing `--release` to package-tarball.sh changes anything.

**Comment by matthiasdiener** [comment]
> Does this work correctly from a packaged tarball too? I get this from CMake with current master from a tarball:
> 
> ```
> fatal: not a git repository (or any of the parent directories): .git
> fatal: not a git repository (or any of the parent directories): .git
> ==============================
> -- Configuring done
> Charm++ 6.10.2 configuration: 
>   OS version:      Ubuntu 18.04.5 LTS (Linux 5.4.0-1031-azure, x86_64)
>   Charm++ version:  (dev)
> ```
> 
> I don't know if passing `--release` to package-tarball.sh changes anything.

Good point, should be fixed now.

**Comment by evan-charmworks** [comment]
buildold says this:

```
fatal: not a git repository (or any of the parent directories): .git
Copying charm-version-git.h.new = "" over charm-version-git.h = 
./system_ln $PWD/charm-version-git.h ../include/charm-version-git.h
```

**Comment by matthiasdiener** [comment]
> buildold says this:
> 
> ```
> fatal: not a git repository (or any of the parent directories): .git
> Copying charm-version-git.h.new = "" over charm-version-git.h = 
> ./system_ln $PWD/charm-version-git.h ../include/charm-version-git.h
> ```

Ok, but this is not a regression with this PR I think.

**Review by evan-charmworks — APPROVED** [review]

File: buildold [hunk 9]
```diff
@@ -761,8 +761,8 @@ EOF
     fi
     $DESTINATION/tmp/system_ln $DESTINATION/$newdir $newdir
   done
-  rm -f VERSION
-  $DESTINATION/tmp/system_ln $DESTINATION/include/VERSION VERSION
+  rm -f charm-version.h
+  $DESTINATION/tmp/system_ln $DESTINATION/include/charm-version.h charm-version.h
```
Comment (minitu) [inline buildold:765]: Do we also want to remove `charm-version-git.h` here?

Comment (matthiasdiener) [reply buildold:765]: Not sure, but I think the `rm` refers to the symlink only, and there is no symlink created for `charm-version-git.h`.

**Review by minitu — APPROVED** [review]

**Pull request merged**